java-checks/src/main/java/org/sonar/java/checks/OSCommandsPathCheck.java

```diff
@@ -0,0 +1,227 @@
+    "./",
+    "../",
+    "~/",
+    "\\",
+    ".\\",
+    "..\\"
+  );
+
+  private static final Pattern WINDOWS_DISK_PATTERN = Pattern.compile("^[A-Z]:\\\\.*");
+
+  private static final String MESSAGE = "Make sure the \"PATH\" used to find this command includes only what you intend.";
+
+  @Override
+  protected MethodMatchers getMethodInvocationMatchers() {
+    return MethodMatchers.or(
+      RUNTIME_EXEC_MATCHER,
+      PROCESS_BUILDER_MATCHER,
+      PROCESS_BUILDER_COMMAND_MATCHER
+    );
+  }
+
+
+  private static boolean isCompliant(String command) {
+    return STARTS.stream().anyMatch(command::startsWith) ||
+      WINDOWS_DISK_PATTERN.matcher(command).matches();
+  }
+
+  @Override
+  protected void onConstructorFound(NewClassTree tree) {
+    Arguments arguments = tree.arguments();
+    if (arguments.isEmpty()) {
+      return;
+    }
+    ExpressionTree argument = arguments.get(0);
+    processArgument(argument);
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree tree) {
+    Arguments arguments = tree.arguments();
+    if (arguments.isEmpty()) {
+      return;
+    }
+    ExpressionTree argument = arguments.get(0);
+    processArgument(argument);
+  }
+
+  private void processArgument(ExpressionTree tree) {
+    ExpressionTree argument = ExpressionUtils.skipParentheses(tree);
+    if (argument.is(Tree.Kind.STRING_LITERAL)) {
+      if (isStringLiteralCommandValid(argument)) {
+        return;
+      }
+      reportIssue(argument, MESSAGE);
+    } else if (argument.is(Tree.Kind.NEW_ARRAY)) {
+      if (isNewArrayCommandValid(argument)) {
+        return;
+      }
+      reportIssue(argument, MESSAGE);
+    } else if (argument.is(Tree.Kind.IDENTIFIER)) {
+      IdentifierTree identifier = (IdentifierTree) argument;
+      if (isIdentifierCommandValid(identifier)) {
+        return;
+      }
+      reportIssue(argument, MESSAGE);
+    } else if (argument.is(Tree.Kind.METHOD_INVOCATION)) {
+      if (isListCommandValid(argument)) {
+        return;
+      }
+      reportIssue(argument, MESSAGE);
+    }
+  }
+
+  private static boolean isStringLiteralCommandValid(ExpressionTree stringLitteral) {
+    Optional<String> command = stringLitteral.asConstant(String.class);
+    return !command.isPresent() || isCompliant(command.get());
+  }
+
+  private static boolean isIdentifierCommandValid(IdentifierTree identifier) {
+    Symbol symbol = identifier.symbol();
+    Type type = symbol.type();
+    if (type.is(STRING_TYPE)) {
+      return isStringLiteralCommandValid(identifier);
+    }
+    //The identifier must be an array or a list
+    Tree declaration = symbol.declaration();
+    if (declaration == null || !declaration.is(Tree.Kind.VARIABLE)) {
+      return true;
+    }
+    VariableTree variable = (VariableTree) declaration;
+    ExpressionTree initializer = variable.initializer();
+    if (initializer == null) {
+      return true;
+    }
+    Tree.Kind initializerKind = initializer.kind();
```

Comment:
As discussed, you can extract this code into a separate method.

Reply:
Done